types/underscore/underscore-tests.ts

```diff
@@ -2691,6 +2745,110 @@ undefinedIdentityIterateeResult; // $ExpectType StringRecord
 
 // Objects
 
+// mapObject
+{
+    // function iteratee - objects
+    _.mapObject(mixedTypeRecord, mixedTypeRecordValueIterator, context); // $ExpectType { a: string; b: string; c: string; }
+    _(mixedTypeRecord).mapObject(mixedTypeRecordValueIterator, context); // $ExpectType { a: string; b: string; c: string; }
+    extractChainTypes(_.chain(mixedTypeRecord).mapObject(mixedTypeRecordValueIterator, context)); // $ExpectType ChainType<{ a: string; b: string; c: string; }, string>
+
+    // function iteratee - dictionaries
+    _.mapObject(stringRecordDictionary, stringRecordDictionaryValueIterator, context); // $ExpectType { [x: string]: string; }
+    _(stringRecordDictionary).mapObject(stringRecordDictionaryValueIterator, context); // $ExpectType { [x: string]: string; }
+    extractChainTypes(_.chain(stringRecordDictionary).mapObject(stringRecordDictionaryValueIterator, context)); // $ExpectType ChainType<{ [x: string]: string; }, string>
+
+    // function iteratee - any
+    _.mapObject(anyValue, stringRecordDictionaryValueIterator, context); // $ExpectType { [x: string]: string; }
+    _(anyValue).mapObject(stringRecordDictionaryValueIterator, context); // $ExpectType { [x: string]: string; }
+    extractChainTypes(_.chain(anyValue).mapObject(stringRecordDictionaryValueIterator, context)); // $ExpectType ChainType<{ [x: string]: string; }, string>
+
+    // partial object iteratee - objects
+    _.mapObject(mixedTypeRecord, partialStringRecord); // $ExpectType { a: boolean; b: boolean; c: boolean; }
+    _(mixedTypeRecord).mapObject(partialStringRecord); // $ExpectType { a: boolean; b: boolean; c: boolean; }
+    extractChainTypes(_.chain(mixedTypeRecord).mapObject(partialStringRecord)); // $ExpectType ChainType<{ a: boolean; b: boolean; c: boolean; }, boolean>
+
+    // partial object iteratee - any
+    _.mapObject(anyValue, partialStringRecord); // $ExpectType { [x: string]: boolean; }
+    _(anyValue).mapObject(partialStringRecord); // $ExpectType { [x: string]: boolean; }
+    extractChainTypes(_.chain(anyValue).mapObject(partialStringRecord)); // $ExpectType ChainType<{ [x: string]: boolean; }, boolean>
+
+    // property name iteratee - objects
+    _.mapObject(mixedTypeRecord, stringRecordProperty); // $ExpectType { a: string; b: any; c: any; }
+    _(mixedTypeRecord).mapObject(stringRecordProperty); // $ExpectType { a: string; b: any; c: any; }
+    const result: ChainType<{ a: string; b: any; c: any; }, any> = extractChainTypes(_.chain(mixedTypeRecord).mapObject(stringRecordProperty));
```

Comment:
Linting with `$ExpectType` was doing the odd thing here I've seen a time or two where if I tell it the expected type should be `ChainType<{ a: string; b: any; c: any; }, any>` it says that the type it arrived at was `ChainType<{ a: string; b: any; c: any; }, never>` and vice versa. Previously I've always ended up making some sort of simplification or other update that has stopped that behavior, but I haven't come up with anything so far this time. This assignment test is a workaround that we can go with if I can't figure anything better out.

Reply:
It doesn't seem like this should be difficult for it to figure out.

1. That part of the return type is `IterateeResult<I, T>`
2. `I` is inferred as `'a'` from the `stringRecordProperty` parameter
3. `T` is known to be `never` from the result of the `_.chain(mixedTypeRecord)` call
4. `'a' extends EnumerableKey` in `IterateeResult<I, T>` should be the first condition that succeeds in `IterateeResult<I, T>`
5. The true branch for that condition yields `any`
6. `IterateeResult<I, T>` in this case should resolve to `any`

Reply:
Oh, `I extends keyof T ? T[I]` is evaluated before `EnumerableKey` and `'a' extends keyof never` is true and results in whatever `never['a']` is 😦

Reply:
I put back an `any` I had removed to make this work.